File: app/contracts/d2/fixtures/csharp-runner/PartialMatch.cs
// 子集深度匹配——语义对齐 ../ts-runner/runner.ts 的 `partialMatch` / ../swift-runner/PartialMatch.swift：
// expected 里出现的每个字段都必须在 actual 里以相等值出现；actual 多出的字段不算失败；数组要求长度相等
// （逐元素再做子集匹配）；expected 完全缺省（C# `null`，对应 TS `undefined`/Swift `Optional.none`）视为
// 『不断言这个字段』，直接通过；expected 显式是 JSON `null`（<see cref="JsonNullMarker.Instance"/>）走
// 严格『actual 也必须显式是 null』比较——`null`（缺失）≠ `JsonNullMarker.Instance`（显式 null），这正是
// Swift 侧 T-048 REWORK #4 第 1 类假阴性修的那处区分（`isExplicitNull`），本文件从第一行就按这个纪律写，
// 不是事后打补丁。
//
// 输入的两侧类型对称——与 Swift 不同，本文件对 actual/expected 两侧统一使用同一个『object? 值域』
// （Dictionary<string,object?>/List<object?>/string/long/double/bool/JsonNullMarker.Instance）：
//   - `expected` 来自 fixture JSON，经 `FixtureDsl.cs` 的 `FixtureJson.ToActual` 从 `JsonElement`
//     解出。
//   - `actual` 是 CSharpFixtureRunner 自己在跑 timeline 过程中，用真实 client 的可观察状态拼出来的
//     同一套值域（构造时遵循同一纪律：显式 JSON null 用 `JsonNullMarker.Instance`，字段不适用/未计算
//     就整个不放进 Dictionary，绝不用裸 `null` 表达『这里其实是 JSON null』——那样会和『键缺失』的
//     C# `null` 语义混淆，见 `JsonNullMarker` 的文档注释）。
//
// C# 与 Swift 的一处真实差异（刻意记录，不是遗漏）：Swift 侧 `PartialMatch.swift` 花了大篇幅处理
// Foundation 把 `Bool` 桥接成 `NSNumber` 之后的假阳性（`NSNumber(value:0) is Bool` 在 Swift 6.3.3
// 上实测为 `true`，见该文件文档注释 #2）——这是 Objective-C/Foundation 运行时特有的『NSNumber 按数值
// 是否恰好是 0/1 做启发式桥接』的历史包袱。C# 没有这个问题：`object` 装箱后的运行时类型精确保留，装箱的
// `bool` 恒为 `System.Boolean`，装箱的 `long`/`int`/`double` 恒为对应的数值类型，`is bool` 模式匹配
// 可靠区分两者，不会把数值 0/1 误判成布尔（也不会反过来）。因此 `ScalarsEqual` 不需要 Swift 那样的
// `objCType` workaround；这是语言运行时差异带来的合理简化，不是抄近路漏掉了 T-048 那一课。

using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpRunner
{
    /// <summary>
    /// 显式 JSON `null` 的哨兵值——与『字段完全缺失』（C# `null`，对应 TS `undefined`/Swift
    /// `Optional.none`）语义不同，必须可区分（PartialMatch 文件头注释）。
    ///
    /// 边界（务必遵守，否则会把这套纪律做反）：这个哨兵只活在 CSharpFixtureRunner 自己的『actual/expected
    /// 统一值域』里，供 `PartialMatch.Match` 消费。任何要喂给真实 `OpenclawGatewayKernelClient`（比如
    /// `ApplyMockResponse`/`ApplyMockEvent` 构造的 wire 帧 `JSONObject`）的字典，如果需要表达 JSON
    /// `null`，一律用裸 C# `null`——那是 `OpenclawWire`/`System.Text.Json` 认得的『JSON null』表达
    /// （`OpenclawWire.ConvertElement` 把 `JsonValueKind.Null` 转成 C# `null`，`JsonObjectExtensions.Get`
    /// / `OpenclawWire.JsonString` 等取值助手也按这个假设工作），绝不能把 `JsonNullMarker.Instance`
    /// 泄漏进那些字典——那样真实 client 的 `is string`/`as string` 转换只会静默拿到 null（不崩溃，但
    /// 语义错了），且 F7 脱敏/PrettyPrint 序列化也认不得这个哨兵类型。
    /// </summary>
    public sealed class JsonNullMarker
    {
        public static readonly JsonNullMarker Instance = new();
        private JsonNullMarker() { }
        public override string ToString() => "null";
    }

    public static class PartialMatch
    {
        public static bool IsExplicitNull(object? v) => v is JsonNullMarker;

        /// <summary>供诊断消息 + `RunFixtureFileAsync` 的 initialState 拒绝消息使用的简易 JSON 风格
        /// 字符串化——不追求转义完整性，只追求可读，镜像 Swift `describeAny`/TS
        /// `JSON.stringify` 在诊断消息里的角色。</summary>
        public static string Stringify(object? v)
        {
            switch (v)
            {
                case null:
                    return "nil";
                case JsonNullMarker:
                    return "null";
                case string s:
                    return $"\"{s}\"";
                case bool b:
                    return b ? "true" : "false";
                case Dictionary<string, object?> dict:
                    return "{" + string.Join(",", dict.Select(kv => $"\"{kv.Key}\":{Stringify(kv.Value)}")) + "}";
                case List<object?> list:
                    return "[" + string.Join(",", list.Select(Stringify)) + "]";
                default:
                    return v.ToString() ?? "nil";
            }
        }

        private static bool TryIntegerValue(object v, out long result)
        {
            switch (v)
            {
                case long l:
                    result = l;
                    return true;
                case int i:
                    result = i;
                    return true;
                case short s:
                    result = s;
                    return true;
                case double d when Math.Truncate(d) == d && Math.Abs(d) < 9.2e18:
                    result = (long)d;
                    return true;
                default:
                    result = 0;
                    return false;
            }
        }

        private static bool TryNumericValue(object v, out double result)
        {
            switch (v)
            {
                case long l:
                    result = l;
                    return true;
                case int i:
                    result = i;
                    return true;
                case double d:
                    result = d;
                    return true;
                default:
                    result = 0;
                    return false;
            }
        }

        private static bool ScalarsEqual(object actual, object expected)
        {
            if (actual is string sa && expected is string sb) return sa == sb;

            // Bool 必须两侧都是 bool 才能比较——C# 装箱运行时类型精确（不像 Swift 需要 objCType
            // workaround，见文件头注释），`is bool` 足够可靠。
            var actualIsBool = actual is bool;
            var expectedIsBool = expected is bool;
            if (actualIsBool || expectedIsBool)
                return actualIsBool && expectedIsBool && (bool)actual == (bool)expected;

            if (TryIntegerValue(actual, out var ia) && TryIntegerValue(expected, out var ib)) return ia == ib;
            if (TryNumericValue(actual, out var na) && TryNumericValue(expected, out var nb)) return na == nb;
            return false;
        }

        /// <summary>`expected`/`actual` 均已从各自来源解出为『统一 object? 值域』（见文件头注释）。</summary>
        public static List<string> Match(object? actual, object? expected, string path)
        {
            var mismatches = new List<string>();

            // C# `null` == TS `expected === undefined` / Swift `Optional.none`（调用方没有提供这个
            // 断言字段）——跳过，不产生任何 mismatch。**区别于**下面的显式 JSON null 分支。
            if (expected is null) return mismatches;

            if (IsExplicitNull(expected))
            {
                if (actual is null || !IsExplicitNull(actual))
                    mismatches.Add($"{path}: 期望 null，实际 {Stringify(actual)}");
                return mismatches;
            }

            if (expected is Dictionary<string, object?> expDict)
            {
                if (actual is not Dictionary<string, object?> actDict)
                {
                    mismatches.Add($"{path}: 期望对象 {Stringify(expected)}，实际 {Stringify(actual)}");
                    return mismatches;
                }
                foreach (var (key, value) in expDict)
                {
                    var actualValue = actDict.TryGetValue(key, out var v) ? v : null;
                    mismatches.AddRange(Match(actualValue, value, $"{path}.{key}"));
                }
                return mismatches;
            }

            if (expected is List<object?> expArr)
            {
                if (actual is not List<object?> actArr)
                {
                    mismatches.Add($"{path}: 期望数组，实际 {Stringify(actual)}");
                    return mismatches;
                }
                if (actArr.Count != expArr.Count)
                {
                    mismatches.Add($"{path}: 期望长度 {expArr.Count}，实际长度 {actArr.Count}");
                    return mismatches;
                }
                for (var i = 0; i < expArr.Count; i++)
                    mismatches.AddRange(Match(actArr[i], expArr[i], $"{path}[{i}]"));
                return mismatches;
            }

            // 标量：actual 缺失（C# null）或本身是显式 null，都不能满足『expected 是某个具体标量』的
            // 比较——TS `actual === expected` 对 undefined/null 与任何非 null/undefined 标量比较恒为
            // false。
            if (actual is null || IsExplicitNull(actual))
            {
                mismatches.Add($"{path}: 期望 {Stringify(expected)}，实际 {Stringify(actual)}");
                return mismatches;
            }
            if (!ScalarsEqual(actual, expected))
                mismatches.Add($"{path}: 期望 {Stringify(expected)}，实际 {Stringify(actual)}");
            return mismatches;
        }
    }
}

File: app/contracts/d2/fixtures/swift-runner/PartialMatch.swift
// 子集深度匹配——语义对齐 ../ts-runner/runner.ts 的 `partialMatch`：expected 里出现的每个字段都必须
// 在 actual 里以相等值出现；actual 多出的字段不算失败；数组要求长度相等（逐元素再做子集匹配）；
// expected 完全缺省（TS 的 `undefined`，Swift 侧对应 Optional.none）视为『不断言这个字段』，直接通过。
//
// T-048 REWORK #4：修正三类此前与 TS 不等价、会把真实 mismatch 放过的假阴性（codex 对抗审复现）：
//   1. 显式 JSON null 与『字段完全缺失』被旧版混为一谈——TS 对 `expected === null` 走严格
//      `actual === expected` 标量比较（`undefined === null` 是 false！），旧版一律放行。现在区分
//      『Optional.none（未提供这个断言字段，对应 TS undefined）』与『isExplicitNull（JSON 字面量
//      null，对应 JSONNull/NSNull 实例）』两种不同情形，只有前者才无条件跳过。
//   2. Foundation 把 `Bool` 桥接成 `NSNumber`后，旧版 `numericValue` 会把它当数字比较，导致
//      `true == 1`、`false == 0` 这类假阳性（TS `===` 对 boolean vs number 严格不等）。**实测发现
//      这个问题比最初判断的更深**：`v is Bool`/`v as? Bool` 本身就不可靠——在这台机器的 Swift
//      6.3.3 上实测 `NSNumber(value: 0) is Bool`/`NSNumber(value: 1) is Bool` 均为 `true`（哪怕
//      这个 NSNumber 明明是从 Codable `Int` 编码来的普通整数、`objCType` 是 `"q"` 不是
//      `"c"`），但 `NSNumber(value: 42) is Bool` 是 `false`——即 Swift 的 NSNumber<->Bool 桥接是
//      **按数值是否恰好等于 0/1 做的启发式判断，不是按类型**，`is Bool` 对『刚好是 0/1 的真整数』
//      同样会给出假阳性（复现：本文件改动后首次跑 basic fixture 的 `evt.message.delta.payload.
//      index:0` 被误报 mismatch）。修法：改查 `objCType`——CFBoolean 桥接来的 NSNumber（真正的
//      JSON `true`/`false`）恒为 `"c"`；经 Codable `Int`/`Int64` 编码的整数（即使数值是 0/1）
//      走 JSON 数字字面量，`objCType` 是 `"q"`，可靠区分（见 `isBoolValue`）。
//   3. 旧版把所有数字统一转 `Double` 比较，大整数经 Double 会丢失精度。现在优先尝试精确的 `Int64`
//      比较（两侧都能无损表示成整数时），只有出现非整数浮点值才落到 Double 比较。
//
// 输入的两侧类型不对称，如实反映两侧数据的来源不同：
//   - `expected` 来自 fixture JSON，经 `JSONAny.value` 解出，是 JSONDecoder 风格的
//     `String`/`Int64`/`Double`/`Bool`/`JSONNull`/`[String: Any]`/`[Any]`。
//   - `actual` 是 SwiftFixtureRunner 自己在跑 timeline 过程中拿真实 client 的可观察状态拼出来的
//     `[String: Any]`/`[[String: Any]]`/`String`/`Int`/`Bool`/`NSNull`——部分经由
//     `JSONSerialization.jsonObject` 往返（`encodeToJSONObject`），是 Foundation 风格的
//     `NSNumber`/`NSNull`，同样是普通 Swift 值，不经过 JSONAny。
// `scalarsEqual`/`integerValue` 负责抹平两侧数值类型的差异（Int vs Int64 vs Double vs NSNumber），
// 同时不引入新的假阳性。

import Foundation

public typealias Mismatch = String

/// 严格意义上的『值本身就是 JSON null 字面量』——不包括『字段完全缺失』（那是 Swift Optional.none，
/// 对应 TS 的 `undefined`，语义不同，见文件头注释 #1）。
func isExplicitNull(_ v: Any) -> Bool {
    if v is NSNull { return true }
    if v is JSONNull { return true }
    return false
}

/// 可靠判断某个值『语义上是不是 JSON 布尔』——**不能只查 `is Bool`/`as? Bool`**（文件头注释 #2
/// 实测记录的更深一层假阳性：Swift 的 NSNumber<->Bool 桥接按数值是否恰好是 0/1 做启发式判断，
/// 对『刚好是 0/1 的真整数』同样会误判为 Bool）。原生 Swift `Bool`（JSONAny/JSONDecoder 解出的
/// expected 一侧）与 CFBoolean 桥接来的 NSNumber（JSONSerialization 解出的 actual 一侧，真正的
/// JSON `true`/`false`）在转成 NSNumber 后 `objCType` 恒为 `"c"`；经 Codable `Int`/`Int64`
/// 编码的整数（即使数值是 0/1）`objCType` 是 `"q"`，可靠区分，不依赖数值本身。
func isBoolValue(_ v: Any) -> Bool {
    if let n = v as? NSNumber {
        return String(cString: n.objCType) == "c"
    }
    return v is Bool
}

/// 精确整数值——只有两侧都能无损表示成 Int64 时才走这条路径，避免大整数经 Double 比较丢失精度
/// （文件头注释 #3）。刻意排除 Bool（`scalarsEqual` 用 `isBoolValue` 在调用本函数之前就已经短路
/// 处理，这里只应处理真正的数字，用同一套 `isBoolValue` 判断排除，不能再用不可靠的 `is Bool`）。
func integerValue(_ v: Any) -> Int64? {
    if isBoolValue(v) { return nil }
    if let i = v as? Int { return Int64(i) }
    if let i64 = v as? Int64 { return i64 }
    if let d = v as? Double, d.truncatingRemainder(dividingBy: 1) == 0, d.magnitude < 0x1p63 {
        return Int64(d)
    }
    if let n = v as? NSNumber {
        // objCType 的 'f'/'d' 是浮点表示——真正的浮点数交给下面的 numericValue/Double 比较处理，
        // 这里只接受整型表示，保留原始精度（不强制先转 Double 再转回来）。
        let objCType = String(cString: n.objCType)
        if objCType == "f" || objCType == "d" { return nil }
        return n.int64Value
    }
    return nil
}

func numericValue(_ v: Any) -> Double? {
    if isBoolValue(v) { return nil }
    if let i = v as? Int { return Double(i) }
    if let i64 = v as? Int64 { return Double(i64) }
    if let d = v as? Double { return d }
    if let n = v as? NSNumber { return n.doubleValue }
    return nil
}

func scalarsEqual(_ a: Any, _ b: Any) -> Bool {
    if let sa = a as? String, let sb = b as? String { return sa == sb }

    // Bool 必须两侧都是 Bool 才能比较——防止 Foundation 的 Bool<->NSNumber 桥接把
    // `true`/`1`、`false`/`0` 误判成相等（文件头注释 #2）。用 `isBoolValue`（基于 objCType），
    // 不用不可靠的 `is Bool`/`as? Bool`。
    let aIsBool = isBoolValue(a)
    let bIsBool = isBoolValue(b)
    if aIsBool || bIsBool {
        guard aIsBool, bIsBool, let ba = a as? Bool, let bb = b as? Bool else { return false }
        return ba == bb
    }

    if let ia = integerValue(a), let ib = integerValue(b) { return ia == ib }
    if let na = numericValue(a), let nb = numericValue(b) { return na == nb }
    return false
}

func describeAny(_ v: Any?) -> String {
    guard let v = v, !isExplicitNull(v) else { return "nil" }
    return "\(v)"
}

/// `expected`/`actual` 均已从各自来源解出为普通 Swift 值（`expected` 一侧调用方负责先做
/// `jsonAny?.value`，本函数不认得 `JSONAny` 本身，只认得它解出来的原始值）。
public func partialMatch(actual: Any?, expected: Any?, path: String) -> [Mismatch] {
    // Optional.none == TS 的 `expected === undefined`（调用方没有提供这个断言字段）——跳过，不
    // 产生任何 mismatch。**区别于**下面的显式 JSON null 分支：JSON 字面量 `null` 解出来是一个具体
    // 的 `JSONNull`/`NSNull` 实例（非 Optional.none），走 isExplicitNull 分支，语义不同。
    guard let expected = expected else { return [] }

    if isExplicitNull(expected) {
        // TS `expected === null` 走严格 `actual === expected` 标量比较：`actual` 必须也是显式
        // null，"字段完全缺失"（`undefined`）不算数——`undefined === null` 在 JS 里是 false，
        // 旧版这里把两者混同是一处真实假阴性（文件头注释 #1）。
        guard let actual = actual, isExplicitNull(actual) else {
            return ["\(path): 期望 null，实际 \(describeAny(actual))"]
        }
        return []
    }

    if let expDict = expected as? [String: Any] {
        guard let actDict = actual as? [String: Any] else {
            return ["\(path): 期望对象 \(describeAny(expected))，实际 \(describeAny(actual))"]
        }
        var out: [Mismatch] = []
        for (key, value) in expDict {
            out += partialMatch(actual: actDict[key], expected: value, path: "\(path).\(key)")
        }
        return out
    }

    if let expArr = expected as? [Any] {
        guard let actArr = actual as? [Any] else {
            return ["\(path): 期望数组，实际 \(describeAny(actual))"]
        }
        guard actArr.count == expArr.count else {
            return ["\(path): 期望长度 \(expArr.count)，实际长度 \(actArr.count)"]
        }
        var out: [Mismatch] = []
        for (index, item) in expArr.enumerated() {
            out += partialMatch(actual: actArr[index], expected: item, path: "\(path)[\(index)]")
        }
        return out
    }

    // 标量：actual 缺失（Optional.none）或本身是显式 null，都不能满足『expected 是某个具体标量』
    // 的比较——TS `actual === expected` 对 undefined/null 与任何非 null/undefined 标量比较恒为 false。
    guard let actual = actual, !isExplicitNull(actual) else {
        return ["\(path): 期望 \(describeAny(expected))，实际 \(describeAny(actual))"]
    }
    if !scalarsEqual(actual, expected) {
        return ["\(path): 期望 \(describeAny(expected))，实际 \(describeAny(actual))"]
    }
    return []
}
